README.md
# pytun

`pytun` is a Python module to manage tun/tap IP tunnel in the same way you would manipulate a file handler.

For now, it is only compatible with Linux, probably Unix, maybe MacOsX, and in the future Windows.

`pytun` is under the MIT license.

## How to use

First of all, clone this repos or use `easy_install` or `pip`.

    pip install pytun
    easy_install pytun

Quite easy. Use the `open()` function.

    import pytun

    tun = pytun.open()  # Open a TUN mode tunnel

For a TAP tunnel, add the `"tap"` argument.

    tun = pytun.open("tap")

`tun` is the handler for the newly created tunnel and is manipulated like a file.

To read/write, use the standard methods `recv([size = 1500])` and `send(buf)`

    buf = tun.recv()
    tun.send(buf)

tun is also `select()` compatible to make your network reactor asynchronous.

    import select

    fds = select.select([tun, ...], [...], [...])

Finally, you can close the tunnel using the method `close()`.

    tun.close()

The tunnel automatically closes when the object is not referenced anymore and the garbage collector destroys the handler.




debian/changelog
pytun (1.0.1-1) stable; urgency=low

  * Initial package build

 -- Harish Kumar <hkumar@d4devops.org>  Mon, 04 Jan 2016 11:18:40 +0000


tests.py
import pytun
import logging
import select

def pprint_buf(buf):
    """ Dirty & convenient function to display the hexademical
        repr. of a buffer.
    """

    DEFAULT_SIZE = 4

    def hex2(i, l = None):
        l = l if l is not None else DEFAULT_SIZE

        h = hex(i).upper()[2:]
        
        if len(h) != l:
            h = "0" * (l - len(h)) + h
        
        return h

    def displayable_char(c):
        if ord(c) < 0x20:
            c = "."

        return c

    print " " * DEFAULT_SIZE,
    for i in range(16): print hex2(i, 2),
    print

    raws = []
    for i, c in enumerate(buf):
        if i % 16 == 0:
            if i:
                print "\t" + "".join(raws)
                raws = []

            print hex2(i),
        raws.append(displayable_char(c))
        
        print hex2(ord(c), 2),

    print "  " * (15 - (i % 16)) + "\t" + "".join(raws)

def main():
    # Configure pytun's logger
    pytun.logger.setLevel(logging.DEBUG)
    logging.basicConfig()

    # Open the tunnel
    try:
        tun = pytun.open()

    except pytun.Tunnel.NotPermitted:
        print
        print "*" * 80
        print "You do have the rights to access the file %s." % (pytun.TUN_KO_PATH, )
        print "Give the access of this file to pytun, or if you trust me,"
        print "elevate this current script to root level."
        print "*" * 80
        print

        raise

    print "*" * 80
    print
    print "OK. The tunnel '%s' had been created." % (tun.name, )
    print
    print "If you want to play with it, first configure it."
    print
    print "1. Set up the network and set an IP"
    print "    $ ifconfig %s 192.168.42.1" % (tun.name, )
    print
    print "2. Add the network route"
    print "    $ route add -net 192.168.42.0/24 dev %s" % (tun.name, )
    print
    print "Then, try to ping some IP in this network ..."
    print "    $ ping 192.168.42.42"
    print
    print "Or do some UDP netcat magic."
    print "    $ nc 192.168.42.42 4242 -u"
    print
    print "Enjoy !"
    print
    print "*" * 80

    try:
        # Receive loop
        while True:
            buf = tun.recv()

            pytun.logger.info("Packet received !")
            pprint_buf(buf)
            print

    except KeyboardInterrupt:
        print "Keyboard interrupt. Closing."
    
    finally:
        # Close the tunnel
        tun.close()


if __name__ == "__main__":
    main()



pytun.py
""" pytun

pytun is a tiny piece of code which gives you the ability to create and
manage tun/tap tunnels on Linux (for now).

"""

__author__ = "Gawen Arab"
__copyright__ = "Copyright 2012, Gawen Arab"
__credits__ = ["Gawen Arab"]
__license__ = "MIT"
__version__ = "1.0.1"
__maintainer__ = "Gawen Arab"
__email__ = "g@wenarab.com"
__status__ = "Beta"

import os
import fcntl
import socket
import struct
import logging
import functools

TUN_KO_PATH = "/dev/net/tun"

logger = logging.getLogger("pytun")

class Tunnel(object):
    """ tun/tap handler class """

    class AlreadyOpened(Exception):
        """ Raised when the user try to open a already-opened
            tunnel. 
        """
        pass

    class NotPermitted(Exception):
        """ Raised when pytun try to setup a new tunnel without
            the good permissions.
        """
        pass
    
    MODES = {
        "tun": 0x0001,
        "tap": 0x0002,
    }
    
    # No packet information flag
    IFF_NO_PI = 0x1000

    # ioctl call
    TUNSETIFF = 0x400454ca
    SIOCSIFHWADDR = 0x8924

    def __init__(self, mode = None, pattern = None, auto_open = None, no_pi = False):
        """ Create a new tun/tap tunnel. Its type is defined by the
            argument 'mode', whose value can be either a string or
            the system value.
            
            The argument 'pattern set the string format used to
            generate the name of the future tunnel. By default, for
            Linux, it is "tun%d" or "tap%d" depending on the mode.
            
            If the argument 'auto_open' is true, this constructor
            will automatically create the tunnel.
            
            If the argument 'no_pi' is true, the device will be
            be opened with teh IFF_NO_PI flag. Otherwise, 4 extra
            bytes are added to the beginning of the packet (2 flag
            bytes and 2 protocol bytes).

        """

        mode = mode if mode is not None else "tun"
        pattern = pattern if pattern is not None else ""
        auto_open = auto_open if auto_open is not None else True

        super(Tunnel, self).__init__()

        self.pattern = pattern
        self.mode = mode
        self.no_pi = self.IFF_NO_PI if no_pi else 0x0000
        
        self.name = None
        self.fd = None

        if isinstance(self.mode, basestring):
            self.mode = self.MODES.get(self.mode, None)

            assert self.mode is not None, "%r is not a valid tunnel type." % (self.mode, )

        if auto_open:
            self.open()

    def __del__(self):
        self.close()

    @property
    def mode_name(self):
        """ Returns the tunnel mode's name, for printing purpose. """

        for name, id in self.MODES.iteritems():
            if id == self.mode:
                return name

    def fileno(self):
        """ Standard function which makes this class 'select()' compatible. """
        return self.fd

    def open(self):
        """ Create the tunnel.
            If the tunnel is already opened, the function will
            raised an AlreadyOpened exception.
        """

        if self.fd is not None:
            raise self.AlreadyOpened()

        logger.debug("Opening %s..." % (TUN_KO_PATH, ))
        self.fd = os.open(TUN_KO_PATH, os.O_RDWR)
        
        logger.debug("Opening %s tunnel '%s'..." % (self.mode_name.upper(), self.pattern, ))
        try:
            ret = fcntl.ioctl(self.fd, self.TUNSETIFF, struct.pack("16sH", self.pattern, self.mode | self.no_pi))

        except IOError, e:
            if e.errno == 1:
                logger.error("Cannot open a %s tunnel because the operation is not permitted." % (self.mode_name.upper(), ))
                raise self.NotPermitted()

            raise

        self.name = ret[:16].strip("\x00")

        logger.info("Tunnel '%s' opened." % (self.name, ))

    def close(self):
        """ Close the tunnel.
            If the tunnel is already closed or never opened,
            do nothing.
        """

        if self.fd is None:
            return
            
        logger.debug("Closing tunnel '%s'..." % (self.name or "", ))

        # Close tun.ko file
        os.close(self.fd)
        self.fd = None

        logger.info("Tunnel '%s' closed." % (self.name or "", ))

    def send(self, buf):
        """ Send the buffer 'buf'. """
        os.write(self.fd, buf)

    def recv(self, size = None):
        """ Receive a buffer. The default size is 1500, the
            classical MTU.
        """

        size = size if size is not None else 1500

        return os.read(self.fd, size)

    def set_mac(self, mac):
        """ Sets the MAC address of the device to 'mac'.
            parameter 'mac' should be a binary representation
            of the MAC address
            Note: Will fail for TUN devices
        """
        mac = map(ord, mac)
        ifreq = struct.pack('16sH6B8', self.name, socket.AF_UNIX, *mac)
        fcntl.ioctl(self.fileno(), self.SIOCSIFHWADDR, ifreq)

    def __repr__(self):
        return "<%s tunnel '%s'>" % (self.mode_name.capitalize(), self.name, )

class TunTunnel(Tunnel):
    """ tun handler class. """
    def __init__(self, *kargs, **kwargs):
        super(TunTunnel, self).__init__("tun", *kargs, **kwargs)

class TapTunnel(Tunnel):
    """ tap handler class. """
    def __init__(self, *kargs, **kwargs):
        super(TapTunnel, self).__init__("tap", **kwargs)

""" Convenient functions to open tunnels. """
tunnel = functools.partial(Tunnel, auto_open = True)
open = tunnel
